### PR TITLE
libfprint: added a fork for Lenovo ThinkPad

### PR DESCRIPTION
**Pull request by cizra**
###### Motivation for this change
Lenovo ThinkPads have a fingerprint sensor with special needs. There's a fork of libfprintd available that supports it.

###### Things done
1. Added `thinkpad` option to `pkgs/development/libraries/libfprint/default.nix`
2. Added `thinkpad` option to `fprintd` (package and service)

Usage:

```
  services.fprintd.package = pkgs.fprintd-thinkpad;
  services.fprintd.enable = true;
```

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/development/libraries/libfprint/default.nix [hunk 1]
```diff
@@ -1,14 +1,19 @@
...
 
 stdenv.mkDerivation rec {
   name = "libfprint-${version}";
   version = "0.99.0";
 
-  src = fetchurl {
-    url = "https://gitlab.freedesktop.org/libfprint/libfprint/uploads/82ba3cef5bdf72997df711eacdb13c0f/libfprint-${version}.tar.xz";
-    sha256 = "16r4nl40y0jri57jiqmdz4s87byblx22lbhyvqpljd6mqm5rg187";
-  };
+  src = fetchurl (
```
Comment (Synthetica9) [inline pkgs/development/libraries/libfprint/default.nix]: You should probably use fetchFromGitHub here

File: pkgs/development/libraries/libfprint/default.nix [hunk 2]
```diff
@@ -1,14 +1,19 @@
-{ stdenv, fetchurl, pkgconfig, meson, ninja, libusb, pixman, glib, nss, gtk3
+{ thinkpad ? false, stdenv, fetchurl, pkgconfig, meson, ninja, libusb, pixman, glib, nss, gtk3
 , coreutils, gtk-doc, docbook_xsl, docbook_xml_dtd_43 }:
 
 stdenv.mkDerivation rec {
   name = "libfprint-${version}";
```
Comment (Synthetica9) [inline pkgs/development/libraries/libfprint/default.nix]:
```suggestion
   name = "libfprint-${version}" + stdenv.lib.optionalString thinkpad "-thinkpad";
```

File: pkgs/development/libraries/libfprint/default.nix [hunk 3]
```diff
@@ -1,14 +1,19 @@
...
   version = "0.99.0";
 
-  src = fetchurl {
-    url = "https://gitlab.freedesktop.org/libfprint/libfprint/uploads/82ba3cef5bdf72997df711eacdb13c0f/libfprint-${version}.tar.xz";
-    sha256 = "16r4nl40y0jri57jiqmdz4s87byblx22lbhyvqpljd6mqm5rg187";
-  };
+  src = fetchurl (
+    if thinkpad then {
+      url = "https://github.com/3v1n0/libfprint/archive/V_0_99_0.tar.gz";  # how to transform 0.99.0 into 0_99_0?
```
Comment (Synthetica9) [inline pkgs/development/libraries/libfprint/default.nix]:
```suggestion
      url = "https://github.com/3v1n0/libfprint/archive/V_${stdenv.lib.replaceStrings ["."] ["_"] version}.tar.gz";
```

**Review by Synthetica9 — CHANGES_REQUESTED** [review]

File: pkgs/development/libraries/libfprint/default.nix (hunk 1 above)
Comment (cizra) [reply pkgs/development/libraries/libfprint/default.nix]: Really? https://github.com/NixOS/nixpkgs/pull/27327#issuecomment-314797154 says I shouldn't.

File: pkgs/development/libraries/libfprint/default.nix (hunk 2 above)
Comment (cizra) [reply pkgs/development/libraries/libfprint/default.nix]: OK

File: pkgs/development/libraries/libfprint/default.nix (hunk 3 above)
Comment (cizra) [reply pkgs/development/libraries/libfprint/default.nix]: not relevant anymore, with `fetchFromGitHub`.

Comment (Synthetica9) [reply pkgs/development/libraries/libfprint/default.nix]: You can still use a tag instead of a commit hash on fetchFromGitHub

**Comment by cizra** [comment]
Please don't merge yet! I'm working out the kinks.

**Comment by cizra** [comment]
@Synthetica9 all your suggestions have been implemented, thanks for these. Are you still requesting other changes?

**Review by Synthetica9 — APPROVED** [review]
LGTM

**Comment by cizra** [comment]
@abbradar it's all yours then!

File: nixos/modules/services/security/fprintd.nix [hunk 4]
```diff
@@ -25,6 +25,16 @@ in
         '';
       };
 
+      package = mkOption {
+        type        = types.package;
```
Comment (worldofpeace) [inline nixos/modules/services/security/fprintd.nix]: Please don't align the `=` vertically. It's not a widely used style or maintained.

File: pkgs/development/libraries/libfprint/default.nix [hunk 5]
```diff
@@ -1,16 +1,28 @@
-{ stdenv, fetchurl, pkgconfig, meson, ninja, libusb, pixman, glib, nss, gtk3
-, coreutils, gtk-doc, docbook_xsl, docbook_xml_dtd_43 }:
+{ thinkpad ? false, stdenv, fetchFromGitHub, fetchurl, pkgconfig, meson, ninja, libusb, pixman, glib, nss, gtk3
+, coreutils, gtk-doc, docbook_xsl, docbook_xml_dtd_43, openssl ? null }:
+
+assert thinkpad -> openssl != null;
 
 stdenv.mkDerivation rec {
-  name = "libfprint-${version}";
+  name = "libfprint-${version}" + stdenv.lib.optionalString thinkpad "-thinkpad";
```
Comment (worldofpeace) [inline pkgs/development/libraries/libfprint/default.nix]: Please use `pname` and reuse it anywhere there's a reference to `libfprint` (not in meta though)

When attributes `pname` and `version` are present they get concatenated with `-` to make `name`.

File: pkgs/tools/security/fprintd/default.nix [hunk 6]
```diff
@@ -1,16 +1,20 @@
-{ stdenv, fetchurl, pkgconfig, intltool
-, libfprint, glib, dbus-glib, polkit, nss, pam, systemd }:
+{ thinkpad ? false
+, stdenv, fetchurl, pkgconfig, intltool, libfprint-thinkpad ? null
+, libfprint ? null, glib, dbus-glib, polkit, nss, pam, systemd }:
 
 stdenv.mkDerivation rec {
-  name = "fprintd-${version}";
+  name = "fprintd-${version}" + stdenv.lib.optionalString thinkpad "-thinkpad";
```
Comment (worldofpeace) [inline pkgs/tools/security/fprintd/default.nix]: `pname` here too.

File: nixos/modules/services/security/fprintd.nix (hunk 4 above)
Comment (cizra) [reply nixos/modules/services/security/fprintd.nix]: fixed

File: pkgs/development/libraries/libfprint/default.nix (hunk 5 above)
Comment (cizra) [reply pkgs/development/libraries/libfprint/default.nix]: fixed

File: pkgs/tools/security/fprintd/default.nix (hunk 6 above)
Comment (cizra) [reply pkgs/tools/security/fprintd/default.nix]: fixed

**Comment by worldofpeace** [comment]
Oh I'm sorry, it appears my suggestion breaks the urls.

This patch should fix that, also remove the `if` for the `src`.

<details>

<summary>patch</summary>


```diff
diff --git a/pkgs/development/libraries/libfprint/default.nix b/pkgs/development/libraries/libfprint/default.nix
index fa6282cec99..fc4a8c7c342 100644
--- a/pkgs/development/libraries/libfprint/default.nix
+++ b/pkgs/development/libraries/libfprint/default.nix
@@ -7,18 +7,19 @@ stdenv.mkDerivation rec {
   pname = "libfprint" + stdenv.lib.optionalString thinkpad "-thinkpad";
   version = "0.99.0";
 
-  src =
-    if thinkpad then
-      fetchFromGitHub {
-        owner = "3v1n0";
-        repo = pname;
-        rev = "2e2e3821717e9042e93a995bdbd3d00f2df0be9c";
-        sha256 = "1vps1wrp7hskf13f7jrv0dwry2fcid76x2w463wplngp63cj7b3b";
-      }
-    else fetchurl {
-      url = "https://gitlab.freedesktop.org/${pname}/${pname}/uploads/82ba3cef5bdf72997df711eacdb13c0f/${pname}-${version}.tar.xz";
+  src = {
+    libfprint-thinkpad = fetchFromGitHub {
+      owner = "3v1n0";
+      repo = "libfprint";
+      rev = "2e2e3821717e9042e93a995bdbd3d00f2df0be9c";
+      sha256 = "1vps1wrp7hskf13f7jrv0dwry2fcid76x2w463wplngp63cj7b3b";
+    };
+
+    libfprint = fetchurl {
+      url = "https://gitlab.freedesktop.org/libfprint/libfprint/uploads/82ba3cef5bdf72997df711eacdb13c0f/libfprint-${version}.tar.xz";
       sha256 = "16r4nl40y0jri57jiqmdz4s87byblx22lbhyvqpljd6mqm5rg187";
     };
+  }.${pname};
 
   buildInputs = [ libusb pixman glib nss gtk3 ]
     ++ stdenv.lib.optional thinkpad openssl;
@@ -28,7 +29,7 @@ stdenv.mkDerivation rec {
   mesonFlags = [ "-Dudev_rules_dir=lib/udev/rules.d" "-Dx11-examples=false" ];
 
   preConfigure = ''
-    substituteInPlace ${pname}/meson.build \
+    substituteInPlace libfprint/meson.build \
       --replace /bin/echo ${coreutils}/bin/echo
   '';
 
diff --git a/pkgs/tools/security/fprintd/default.nix b/pkgs/tools/security/fprintd/default.nix
index 2297b8cdca4..b43be1a6924 100644
--- a/pkgs/tools/security/fprintd/default.nix
+++ b/pkgs/tools/security/fprintd/default.nix
@@ -7,7 +7,7 @@ stdenv.mkDerivation rec {
   version = "0.8.1";
 
   src = fetchurl {
-    url = "https://gitlab.freedesktop.org/libfprint/${pname}/uploads/bdd9f91909f535368b7c21f72311704a/${pname}-${version}.tar.xz";
+    url = "https://gitlab.freedesktop.org/libfprint/fprintd/uploads/bdd9f91909f535368b7c21f72311704a/fprintd-${version}.tar.xz";
     sha256 = "124s0g9syvglgsmqnavp2a8c0zcq8cyaph8p8iyvbla11vfizs9l";
   };
 
``` 

</details>

**Comment by cizra** [comment]
@worldofpeace Are Nix expressions lazily evaluated? If not, you're fetching both sources instead of only the right one.

Also, why'd the URLs break?

**Comment by Synthetica9** [comment]
> Are Nix expressions lazily evaluated?

They are :)

**Comment by cizra** [comment]
@worldofpeace should be better now.

**Comment by worldofpeace** [comment]
@GrahamcOfBorg build libfprint fprintd libfprint-thinkpad fprintd-thinkpad

**Comment by worldofpeace** [comment]
Congrats @cizra :sparkles: Thank you for contributing. 
Also, smart to followup in irc for questions, most nixos devs hang around there.

**Pull request merged**